New `one_hot_encoding_view((query),column[,column,...]` table function for machine learning

vzakaznikov opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
New `one_hot_encoding_view((query),column[,column,...]` table function for machine learning.


Detailed description / Documentation draft:
Adding new `one_hot_encoding_view((query),column[,column,...]` table function.

This table function encodes one or more columns from a query using https://en.wikipedia.org/wiki/One-hot.
The use case is to provide training data for machine learning (https://en.wikipedia.org/wiki/One-hot#Machine_learning_and_statistics) right out of ClickHouse without extra processing.

```sql
user-host :) DROP TABLE IF EXISTS one_hot_source
user-host :) CREATE TABLE one_hot_source (Id String, X String) Engine=Memory
user-host :) INSERT INTO TABLE one_hot_source VALUES (1, 'A'),(2,'B'),(3,'C'),(4,'A'),(5,'D'),(6,'B')
user-host :) SELECT * FROM one_hot_encoding_view((SELECT * FROM one_hot_source),X)
user-host :) SELECT * FROM one_hot_encoding_view((SELECT * FROM one_hot_source),X)

SELECT *
FROM one_hot_encoding_view(
    (
        SELECT *
        FROM one_hot_source
    ), X
)

Query id: 91807e08-3eb3-4767-bffa-44d154b20b8d

┌─Id─┬─X─┬─X.encoded─┐
│ 1  │ A │ [1,0,0,0] │
│ 2  │ B │ [0,1,0,0] │
│ 3  │ C │ [0,0,1,0] │
│ 4  │ A │ [1,0,0,0] │
│ 5  │ D │ [0,0,0,1] │
│ 6  │ B │ [0,1,0,0] │
└────┴───┴───────────┘

6 rows in set. Elapsed: 0.012 sec. 

user-host :) SELECT * FROM one_hot_encoding_view((SELECT * FROM one_hot_source),X,Id)

SELECT *
FROM one_hot_encoding_view(
    (
        SELECT *
        FROM one_hot_source
    ), X, Id
)

Query id: 714cedbf-ca98-438a-97e0-7222af700c2d

┌─Id─┬─X─┬─X.encoded─┬─Id.encoded────┐
│ 1  │ A │ [1,0,0,0] │ [1,0,0,0,0,0] │
│ 2  │ B │ [0,1,0,0] │ [0,1,0,0,0,0] │
│ 3  │ C │ [0,0,1,0] │ [0,0,1,0,0,0] │
│ 4  │ A │ [1,0,0,0] │ [0,0,0,1,0,0] │
│ 5  │ D │ [0,0,0,1] │ [0,0,0,0,1,0] │
│ 6  │ B │ [0,1,0,0] │ [0,0,0,0,0,1] │
└────┴───┴───────────┴───────────────┘

6 rows in set. Elapsed: 0.146 sec. 

user-host :) 
```

The table function works similar to `view` table function but wraps the original query to add encoded columns for each
column to be encoded. 

**Performance**

> One-hot encoding generates a lot of data. Right now `max_block_size` must be set to some low number
> when cardinality of the column is very large.

**10,000x10,000** (`max_block_size=5`)
```
user-host :) SELECT * FROM one_hot_encoding_view((SELECT number FROM numbers(1,10000)),number) FORMAT Null

SELECT *
FROM one_hot_encoding_view(
    (
        SELECT number
        FROM numbers(1, 10000)
    ), number
)
FORMAT `Null`

Query id: a5ed730b-7e49-4a14-8f93-d96e80988b2c

Ok.

0 rows in set. Elapsed: 0.409 sec. Processed 10.00 thousand rows, 80.00 KB (24.45 thousand rows/s., 195.57 KB/s.)
```

**100,000x100,000** (`max_block_size=2`)
```
user-host :) SELECT * FROM one_hot_encoding_view((SELECT number FROM numbers(1,100000)),number) FORMAT Null

SELECT *
FROM one_hot_encoding_view(
    (
        SELECT number
        FROM numbers(1, 100000)
    ), number
)
FORMAT `Null`

Query id: 33cd84c3-1c20-4dba-b37e-6f870f261faa

Ok.

0 rows in set. Elapsed: 50.867 sec. Processed 100.00 thousand rows, 800.00 KB (1.97 thousand rows/s., 15.73 KB/s.)

user-host :) 
```

**1,000,000x1,000,000** (`max_block_size=2`)
```
user-host :) SELECT * FROM one_hot_encoding_view((SELECT number FROM numbers(1,1000000)),number) FORMAT Null

SELECT *
FROM one_hot_encoding_view(
    (
        SELECT number
        FROM numbers(1, 1000000)
    ), number
)
FORMAT `Null`

Query id: 348a94bd-d482-4596-b338-221aa14b35ce

Cancelling query.
Ok.
Query was cancelled.

0 rows in set. Elapsed: 45.968 sec. Processed 8.03 thousand rows, 64.21 KB (174.60 rows/s., 1.40 KB/s.)

```

**OnTime Dataset (small chunk)**

```
user-host :) SELECT * FROM one_hot_encoding_view((SELECT Dest FROM ontime), Dest)
...
4788862 rows in set. Elapsed: 4.238 sec. Processed 4.79 million rows, 23.95 MB (1.13 million rows/s., 5.65 MB/s.)
```

```
user-host :) SELECT * FROM one_hot_encoding_view((SELECT * FROM ontime), Reporting_Airline) FORMAT Null

SELECT *
FROM one_hot_encoding_view(
    (
        SELECT *
        FROM ontime
    ), Reporting_Airline
)
FORMAT `Null`

0 rows in set. Elapsed: 0.862 sec. Processed 4.79 million rows, 3.52 GB (5.55 million rows/s., 4.08 GB/s.)
```

```
user-host :) SELECT * FROM one_hot_encoding_view((SELECT * FROM ontime), Reporting_Airline, Dest) FORMAT Null

SELECT *
FROM one_hot_encoding_view(
    (
        SELECT *
        FROM ontime
    ), Reporting_Airline, Dest
)
FORMAT `Null`

0 rows in set. Elapsed: 3.218 sec. Processed 4.79 million rows, 3.52 GB (1.49 million rows/s., 1.09 GB/s.)

user-host :) 
```

**To do**

* [x] Fix `one_hot_encoding_view` output formatter to include columns.
* [x] fix `EXPLAIN` query (can't be fixed)
* [x] consider moving to using arrays instead of separate columns for each distinct value
* [x] not needed we check column names against the query
* [x] fix hard coded global `max_block_size`. Need a way to pass this to StorageView
* [x] fix SQL injection by using AST visitor to replace identifiers and aliases

vzakaznikov committed:
Adding support for `one_hot_encoding_view((query),column[,column,...]) table function.

vzakaznikov committed:
Converting one hot encoded columns to be UInt8 instead of UInt64

vzakaznikov commented:
Without using this view here are a few different options.

**Create an array**
```sql
WITH
    (
        SELECT groupArray(X) AS mapping
        FROM
        (
            SELECT DISTINCT X
            FROM one_hot_source
            ORDER BY X ASC
        )
    ) AS mapping,
    arrayResize([0], length(mapping)) AS _encoding
SELECT
    Id,
    X,
    arrayMap((x, i) -> ((mapping[i]) = X), _encoding, arrayEnumerate(mapping)) AS encoding
FROM one_hot_source
FORMAT CSVWithNames
```
```bash
Query id: 6fba55fd-a823-48f8-9e31-9e39879ad130
"Id","X","encoding"
"1","A","[1,0,0,0]"
"2","B","[0,1,0,0]"
"3","C","[0,0,1,0]"
"4","A","[1,0,0,0]"
"5","D","[0,0,0,1]"
"6","B","[0,1,0,0]"
```
Other options
```sql
SELECT 'version1';
-- version 1
-- Naive solution
SELECT 
   Id, 
   X,
   equals(X, 'A') as A,
   equals(X, 'B') as B,
   equals(X, 'C') as C,
   equals(X, 'D') as D
FROM one_hot_source FORMAT CSVWithNames;
​
SELECT 'version2';
-- version 2
-- Optimized solution that moves comparison to be
-- only between integers of the UInt64 type
--
-- NOTE: another solution is to use toLowCardinality
-- data type for the feature column and use version 1
-- of the query.
WITH (
   SELECT 
        groupArray(X) AS mapping
   FROM (SELECT DISTINCT X FROM one_hot_source ORDER BY X)
) AS mapping
SELECT
   Id,
   X,
   equals(indexOf(mapping,X), 1) AS A,
   equals(indexOf(mapping,X), 2) AS B,
   equals(indexOf(mapping,X), 3) AS C,
   equals(indexOf(mapping,X), 4) AS D
FROM one_hot_source FORMAT CSVWithNames;
```

vzakaznikov committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into one_hot_encoding_table_function

vzakaznikov committed:
Fix issue in QueryNormalizer.cpp due to one_hot_encoding_view
Fix typo in TableFunctionOneHotEncodingView.h
Fix output formatting ASTFunction.cpp

vzakaznikov committed:
Fixing clang error.

filimonov commented:
Hacky, but interesting (@vzakaznikov did it on weekend for some pet project).

I would actually generalize it a bit to something like `pivoting_view` or `pivoting_transform`.

FArthur-cmd committed:
Update 01958_one_hot_encoding_view.sql

FArthur-cmd commented:
There is one big problem in this realisation. In `EXPLAIN` queries subqueries  will be executed. You can test this fact with

`EXPLAIN SELECT * FROM one_hot_encoding_view((SELECT toString(number) as num FROM numbers(1000000000) WHERE number > 999999990), num);`

Its execution time is equal to execution of 

`SELECT toString(number) as num FROM numbers(1000000000) WHERE number > 999999990)`.

We need to correct this fact. Right now, It's not clear how to fix it.

FArthur-cmd reviewed (changes requested):
Need to fix `explain` query,

vzakaznikov commented:
We also need to fix all the issues pointed out by @alexey-milovidov as related to code in 
https://github.com/ClickHouse/ClickHouse/blob/013e66a0336bd7adb0d31429a997a5a65c348f04/src/TableFunctions/TableFunctionOneHotEncodingView.cpp#L24

> English: There is an SQL injection in this piece of code - escaping for column and table names is not taken into account. 
> It is not checked that base_query_str is a request... It is better to make it in the form of an AST. And to throw base_query_str in the form of an AST, without serializing it to a string.
>
> executeQuery should be used not the one that outputs bytes, but the one that outputs the request pipeline (BlockIO).
>
> STYLE_CHECK_ALLOW_STD_STRING_STREAM - this check exists for a reason.
>
> The code is technically incorrect, since the format may not be TSV, and the TSV format can escape the line feed as \ and the > line feed (although we escape as \n for convenience when output).
>
> When using getline (which cannot be used), the status is not checked, and in some cases an error exception will be thrown.
>
> Strings - very slow.

vzakaznikov commented:
> Need to fix `explain` query,

Thanks @FArthur-cmd! I'll look into fixing the `explain` query as well as addressing comments from Alexey. 

I might change the implementation to just return an encoded array instead of breaking each value into its own column
as it does not look like having even 10s of thousands of columns is scalable when feature column has high cardinality. 

Also, the advantage of moving to arrays is that if we use something like the query below for the wrapper then we don't need to even send a separate query to get the distinct values and that removes all the bugs pointed out by Alexey.

```sql
WITH
    (
        SELECT groupArray(toLowCardinality(X)) AS mapping
        FROM
        (
            SELECT DISTINCT X
            FROM one_hot_source
            ORDER BY X ASC
        )
    ) AS mapping,
    arrayResize([0], length(mapping)) AS _encoding
SELECT
    Id,
    toLowCardinality(X) AS X,
    arrayMap((x, i) -> ((mapping[i]) = X), _encoding, arrayEnumerate(mapping)) AS encoding
FROM one_hot_source
FORMAT CSVWithNames
```

vzakaznikov commented:
Need to also fix
> 2021-08-03 15:41:19 /build/obj-x86_64-linux-gnu/../src/TableFunctions/TableFunctionOneHotEncodingView.cpp:24:42: error: method 'getColumnDistinctValues' can be made static [readability-convert-member-functions-to-static,-warnings-as-errors]

vzakaznikov commented:
> Hacky, but interesting (@vzakaznikov did it on weekend for some pet project).
> 
> I would actually generalize it a bit to something like `pivoting_view` or `pivoting_transform`.

I have looked into generalizing this approach using something like `pivoting_view` it will not work out that cleanly for 
one-hot encoding and I think the function will be every more _hackier_ that it is.

The considered syntax for something like `pivot_view` was

```sql
pivot_view((select query),(subquery to get pivots), pivot_expression,[pivot_column_name])  

SELECT * FROM pivot_view(
    (
         SELECT * 
         FROM one_hot_source
    ),
    (
         SELECT DISTINCT X 
         FROM one_hot_source
    ), 
    (
         equals(X, {pivot_value:String})
    ), 
    (
         X.{pivot_value:String}
    )
)
```

If we had pivots they definitely could be used but only if pivot values can be specified using a sub-query and there is a way to add prefixes to pivot column names.

A non-SQL compliant syntax that would fit nicely for this problem would be something like the below but would require
its implementation in our SQL-language.
```
WITH (SELECT DISTINCT X FROM one_hot_source ORDER BY X) as pivots
SELECT 
   Id,
   X, 
   FOR pivot_value IN pivots (equals(X, pivot_value))) as pivot_value
FROM one_hot_source FORMAT CSVWithNames;
```
where syntax for the `FOR expression` is something like

`FOR value IN (subquery_or_array_or_with_clause_value) (expression) AS value`

vzakaznikov commented:
@alexey-milovidov, do you think it is crazy to consider for our SQL-language to have something like the `FOR expression`
I describe in https://github.com/ClickHouse/ClickHouse/pull/27067#issuecomment-892203429?

vzakaznikov committed:
Moving away from generating column per each distinct value of the feature
and instead moved to using one Array(UInt8) column.

vzakaznikov committed:
Merge branch 'one_hot_encoding_table_function' of https://github.com/vzakaznikov/ClickHouse into one_hot_encoding_table_function

vzakaznikov commented:
> Need to fix `explain` query,

Should be fixed now. I have changed the implementation to use Array(UInt8) column.

vzakaznikov commented:
> > Need to fix `explain` query,
> 
> Should be fixed now. I have changed the implementation to use Array(UInt8) column.

Actually, not yet fixed :(. I still need to see why `EXPLAIN` is executing the query.

FArthur-cmd commented:
Do you know what exactly is being executing? Maybe try to add output (or better print stack trace) in some sections, to see what part is executing (or use gdb for this purposes). Without stack trace of execution it will be difficult to say what exactly goes wrong.

vzakaznikov commented:
> Do you know what exactly is being executing? Maybe try to add output (or better print stack trace) in some sections, to see what part is executing (or use gdb for this purposes). Without stack trace of execution it will be difficult to say what exactly goes wrong.

@FArthur-cmd , the same bug is present in the `view()` table function.

```
user-host :) EXPLAIN SELECT * FROM view(SELECT * FROM system.numbers LIMIT 300) FORMAT Null

EXPLAIN
SELECT *
FROM view(
    SELECT *
    FROM system.numbers
    LIMIT 300
)
FORMAT `Null`

Query id: 46a3b344-6cfb-46e1-8e1e-98dcb4807936

Ok.

0 rows in set. Elapsed: 0.003 sec. 

user-host :) 
```

vzakaznikov commented:
> Do you know what exactly is being executing? Maybe try to add output (or better print stack trace) in some sections, to see what part is executing (or use gdb for this purposes). Without stack trace of execution it will be difficult to say what exactly goes wrong.

@FArthur-cmd, the same bug with `EXPLAIN` is present in `view()` table function.

```
user-host :) EXPLAIN SELECT * FROM view(WITH (SELECT groupArray(`number`) FROM (SELECT DISTINCT `number` FROM (SELECT * FROM system.numbers LIMIT 200000000) ORDER BY `number`)) AS `__ohe_mapping_number`, arrayEnumerate(`__ohe_mapping_number`) AS `__ohe_mapping_idx_number` SELECT *,arrayMap(i -> `__ohe_mapping_number`[i] = `number`,`__ohe_mapping_idx_number`) AS `number.encoded` FROM (SELECT * FROM system.numbers LIMIT 200000000))

EXPLAIN
SELECT *
FROM view(
    WITH
        (
            SELECT groupArray(number)
            FROM
            (
                SELECT DISTINCT number
                FROM
                (
                    SELECT *
                    FROM system.numbers
                    LIMIT 200000000
                )
                ORDER BY number ASC
            )
        ) AS __ohe_mapping_number,
        arrayEnumerate(__ohe_mapping_number) AS __ohe_mapping_idx_number
    SELECT
        *,
        arrayMap(i -> ((__ohe_mapping_number[i]) = number), __ohe_mapping_idx_number) AS `number.encoded`
    FROM
    (
        SELECT *
        FROM system.numbers
        LIMIT 200000000
    )
)

Query id: dede47d6-f23d-49db-98a0-21bdcd08f154
```

vzakaznikov commented:
@FArthur-cmd, actually, even without `view()` table function it gets stuck. Can you try on your side?

```
user-host :) select version()

SELECT version()

Query id: 23475f4a-2edb-45a6-8c69-1a5e53f28a17

┌─version()─┐
│ 21.9.1.1  │
└───────────┘

1 rows in set. Elapsed: 0.002 sec. 

user-host :) 
```

```
EXPLAIN
WITH
    (
        SELECT groupArray(number)
        FROM
        (
            SELECT DISTINCT number
            FROM
            (
                SELECT *
                FROM system.numbers
                LIMIT 200000000
            )
            ORDER BY number ASC
        )
    ) AS __ohe_mapping_number,
    arrayEnumerate(__ohe_mapping_number) AS __ohe_mapping_idx_number
SELECT
    *,
    arrayMap(i -> ((__ohe_mapping_number[i]) = number), __ohe_mapping_idx_number) AS `number.encoded`
FROM
(
    SELECT *
    FROM system.numbers
    LIMIT 200000000
)

Query id: c95e9594-2140-4dee-a565-7b6edca03307
```

vzakaznikov commented:
@FArthur-cmd, I have filed https://github.com/ClickHouse/ClickHouse/issues/27189 for the issue with `EXPLAIN`.

vzakaznikov committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into one_hot_encoding_table_function

FArthur-cmd commented:
It's not stuck.I 'll try to find out about `view`

vzakaznikov committed:
Improving array based implementation.

vzakaznikov commented:
@FArthur-cmd, it looks like that `view()` table function can't pass query settings or session settings to the underlying `StorageView` so for example `set max_block_size = 1` will be ignored. Do you know how to fix that?

vzakaznikov committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into one_hot_encoding_table_function

vzakaznikov committed:
Adding support to turn off using global context in StorageView.

vzakaznikov committed:
Fixing SQL injections by verifying all the column arguments.

vzakaznikov commented:
> @FArthur-cmd, it looks like that `view()` table function can't pass query settings or session settings to the underlying `StorageView` so for example `set max_block_size = 1` will be ignored. Do you know how to fix that?

I has to add support to StorageView to not use global context when not needed (https://github.com/ClickHouse/ClickHouse/blob/5202a60017a622dc2dc5d7452f2576776f179554/src/Storages/StorageView.cpp#L92).

vzakaznikov commented:
> Need to fix `explain` query,

@FArthur-cmd, there is fundamentally no way to me fix the `EXPLAIN` query. The `EXPLAIN` behaves in the same way if the internal query is executed without this view so there is nothing that can be done in this code.

vzakaznikov commented:
@FArthur-cmd, this PR is ready for your review. SQL injection was fixed by making sure that all columns passed by the user match the specified query table structure.

FArthur-cmd commented:
@vzakaznikov 
> it looks like that view() table function can't pass query settings or session settings to the underlying StorageView so for example set max_block_size = 1 will be ignored. Do you know how to fix that?
It is not a mistake, it is a fix for #24095. There were some categories of queries, which need this fix. (We will think of this problem)

> there is fundamentally no way to me fix the EXPLAIN query. The EXPLAIN behaves in the same way if the internal query is executed without this view so there is nothing that can be done in this code.

Let's forget about this problem for some time. This problem is connected with internal queries, which you are using. If you can avoid using them that will be great. Otherwise, just pass this problem.

FArthur-cmd commented on src/TableFunctions/TableFunctionOneHotEncodingView.h:
```diff
@@ -0,0 +1,36 @@
+    static constexpr auto name = "one_hot_encoding_view";
+    std::string getName() const override { return name; }
+private:
+    std::tuple<const String, const String> getWithAndSelectForColumn(String base_query_str, const String & column_name);
```
Static?

vzakaznikov replied:
Yes. I will make it static.

vzakaznikov replied:
With the latest changes this method uses object attributes so it does not have to be static now.

FArthur-cmd reviewed (changes requested):
Still have problem with this method

FArthur-cmd commented:
@vzakaznikov 
>it looks like that view() table function can't pass query settings or session settings to the underlying StorageView so for example set max_block_size = 1 will be ignored. Do you know how to fix that?

We have #27227. Will it help you?

vzakaznikov committed:
Fixing SQL injection using AST visitors in TableFunctionOneHotEncodingView.

vzakaznikov commented:
@FArthur-cmd, please see the commit that should fix SQL injections in the generated query. Please review this PR again when you have a chance. Thanks!

vzakaznikov committed:
Updated src/Interpreters/ya.make

vzakaznikov committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into one_hot_encoding_table_function

vzakaznikov committed:
Updating tests/queries/0_stateless/01958_one_hot_encoding_view.sql

hodgesrm commented:
I really like the idea of putting one hot encoding (1HE) support into ClickHouse. That said I have some questions about the view approach. 

1.) It seems as if this approach requires recalculating 1HE mappings for every query, potentially over very large datasets. This looks like a resource-intensive approach for what amounts to an aggregation problem. Normally we would put such aggregates into a mat view, then perform a merge to get the final mapping of values to 1HE.  I don't see how this path enables that kind of efficiency. 

2.) Does it matter for ML analyses if 1HE values change over time for the same column values? For example, if you are pulling multipble training sets out of source data with on-going inserts, the mapping from source value to 1HE can change between queries. That seems as if it would mess up algorithms like random forests. 

It seems to me that another approach would be to aggregate 1HE mappings as a first step, then join on them, for example at the initiator node in a distributed query. That could avoid dragging a huge amount of data around in the initial scan as well as shipping it over the network.  In other words, why not have a one_hot_encoding() function that works analogously to uniq()? 

p.s., I'm not an expert on ML or ClickHouse aggregation but it seems to me there's a ClickHouse way to solve this problem.

vzakaznikov commented:
> 1.) It seems as if this approach requires recalculating 1HE mappings for every query, potentially over very large datasets. This looks like a resource-intensive approach for what amounts to an aggregation problem. Normally we would put such aggregates into a mat view, then perform a merge to get the final mapping of values to 1HE. I don't see how this path enables that kind of 
efficiency.

The `one_hot_encoding_view` table function can be used as the base query to create MV to get the efficiency of not recalculating. The only assumption is that column cardinality shall remain static. The final results then can be queried from the MV.

> 2.) Does it matter for ML analyses if 1HE values change over time for the same column values? For example, if you are pulling multipble training sets out of source data with on-going inserts, the mapping from source value to 1HE can change between
queries. That seems as if it would mess up algorithms like random forests.

Again, current implementation does not support dynamic changes in the cardinality of the encoded columns either adding or removing new distinct values.

--

However, I think we can do better and store this 1-hot encoding just as an offset (position) into the bit mask. Bitmaps (https://clickhouse.tech/docs/en/sql-reference/functions/bitmap-functions/)
seems like to the something that could if used but currently lack a way to get a position of the value in the bitmap.
I am not sure if bitmaps could support `bitmapPos(bitmap, element)` function to get a position of the element inside the bitmap
but I am not sure how these bitmaps are implemented internally. Maybe @andyyzh could comment if something like `bitmapPos()` can be implemented for bitmaps.

hodgesrm commented:
@vzakaznikov, can you show an example of a materialized view definition and the corresponding select out of it? I'm unclear what the partially-aggregated state would look like as well as the merge to complete aggregation. It looks to me as if you would hit the cardinality issue if you computed 1HE on inserted blocks, since you would not see all values. 

Thanks!

vzakaznikov commented:
> @vzakaznikov, can you show an example of a materialized view definition and the corresponding select out of it? I'm unclear what the partially-aggregated state would look like as well as the merge to complete aggregation. It looks to me as if you would hit the cardinality issue if you computed 1HE on inserted blocks, since you would not see all values.
> 
> Thanks!

You are right! We can't create MV right now

FArthur-cmd committed:
Update TableFunctionOneHotEncodingView.h

FArthur-cmd committed:
Update TableFunctionOneHotEncodingView.cpp

FArthur-cmd commented on tests/queries/0_stateless/01958_one_hot_encoding_view.sql:
```diff
@@ -0,0 +1,17 @@
+SELECT * FROM one_hot_encoding_view((SELECT 1),`SHOW TABLES`); -- {serverError 36}
+-- check dummy base query with valid colum
+SELECT * FROM one_hot_encoding_view((SELECT 1),`1`);
+-- check that we can apply max_block_size
```
It seems, like this part of test leads to memory or time limit. Can you fix this? (probably we don't need to check that settings work?)

alexey-milovidov commented on src/TableFunctions/TableFunctionOneHotEncodingView.cpp:
```diff
@@ -0,0 +1,140 @@
+    String with;
+    String select;
+
+    with = "(SELECT groupArray(`" + ohe_column_name +"`) FROM ";
```
We cannot merge this code.

vzakaznikov replied:
> The implementation looks very dirty. Maybe take some time to reconsider this feature or look for alternative?

This feature was just an experiment. I will close this PR.

vzakaznikov replied:
> We cannot merge this code.

While it might not be pretty, I don't believe this code has any SQL injection issues. The choice of not defining query in AST was deliberate due to impracticality of defining complex query using an AST tree.

alexey-milovidov commented on tests/queries/0_stateless/01958_one_hot_encoding_view.sql:
```diff
@@ -0,0 +1,17 @@
+SELECT * FROM one_hot_encoding_view((),`SHOW TABLES`); -- { clientError 62} 
+-- check dummy base query
+SELECT * FROM one_hot_encoding_view((SELECT 1),`SHOW TABLES`); -- {serverError 36}
+-- check dummy base query with valid colum
```
Typo.

alexey-milovidov commented:
The implementation looks very dirty. Maybe take some time to reconsider this feature or look for alternative?

vzakaznikov commented:
This feature needs to be implemented as an aggregate function similar to `uniqExact` (https://clickhouse.tech/docs/en/sql-reference/aggregate-functions/reference/uniqexact/) and not as a view table function. Nonetheless, it does show an example how a convenience view table function can be implemented to hide a complex query from the user.
